path_login.go

```diff
@@ -0,0 +1,239 @@
+			Issuer:   config.BoundIssuer,
+			Subject:  role.BoundSubject,
+			ID:       role.BoundID,
+			Audience: jwt.Audience(role.BoundAudiences),
+			Time:     time.Now(),
+		}
+
+		if err := claims.Validate(expected); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
+		}
+
+	case config.OIDCIssuerURL != "":
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		verifier := provider.Verifier(&oidc.Config{
+			SkipClientIDCheck: true,
+		})
+
+		idToken, err := verifier.Verify(ctx, token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating signature: {{err}}", err).Error()), nil
+		}
+
+		if err := idToken.Claims(allClaims); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("unable to successfully parse all claims from token: {{err}}", err).Error()), nil
+		}
+
+		var jtiVal string
+		if jti, ok := allClaims["jti"]; ok {
+			if jtiStr, ok := jti.(string); ok {
+				jtiVal = jtiStr
+			}
+		}
+
+		switch {
+		case config.BoundIssuer != "" && config.BoundIssuer != idToken.Issuer:
+			return logical.ErrorResponse("iss claim does not match bound issuer"), nil
+		case role.BoundSubject != "" && role.BoundSubject != idToken.Subject:
+			return logical.ErrorResponse("sub claim does not match bound subject"), nil
+		case len(role.BoundAudiences) != 0 && !strutil.StrListSubset(role.BoundAudiences, idToken.Audience):
+			return logical.ErrorResponse("aud claim does not match any bound audience"), nil
+		case role.BoundID != "" && role.BoundID != jtiVal:
+			return logical.ErrorResponse("jti claim does not match bound id"), nil
+		}
+
+	default:
+		return nil, errors.New("unhandled case during login")
+	}
+
+	userClaimRaw, ok := allClaims[role.UserClaim]
+	if !ok {
+		return logical.ErrorResponse(fmt.Sprintf("%q claim not found in token", role.UserClaim)), nil
+	}
+	userName, ok := userClaimRaw.(string)
+	if !ok {
+		return logical.ErrorResponse(fmt.Sprintf("%q claim could not be converted to string", role.UserClaim)), nil
+	}
+
+	var groupAliases []*logical.Alias
+	if role.GroupsClaim != "" {
+		groupsClaimRaw, ok := allClaims[role.GroupsClaim]
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("%q claim not found in token", role.GroupsClaim)), nil
+		}
+		groups, ok := groupsClaimRaw.([]interface{})
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("%q claim could not be converted to string list", role.GroupsClaim)), nil
+		}
+		for _, groupRaw := range groups {
+			group, ok := groupRaw.(string)
+			if !ok {
+				return logical.ErrorResponse(fmt.Sprintf("value %v in groups claim could not be parsed as string", groupRaw)), nil
+			}
+			if group == "" {
+				continue
+			}
+			groupAliases = append(groupAliases, &logical.Alias{
+				Name: group,
+			})
+		}
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
```

Comment:
I believe you need to add the BoundCIDRs to the auth block.

Reply:
Ah, right.

Reply:
Done.